### PR TITLE
Spark: Split SparkCatalogTestBase and allow running tests with one catalog

### PR DESCRIPTION
**Pull request by hililiwei**
https://github.com/apache/iceberg/issues/3499

1. new enum named SparkCatalogType with three test spark catalog 'testhive'\'testhadoop'\'spark_catalog''

```
public enum SparkCatalogType {
  TEST_HIVE("testhive", SparkCatalog.class.getName(), ImmutableMap.of(
      "type", "hive",
      "default-namespace", "default"
  )),
  TEST_HADOOP("testhadoop", SparkCatalog.class.getName(), ImmutableMap.of(
      "type", "hadoop"
  )),
  SPARK_CATALOG("spark_catalog", SparkSessionCatalog.class.getName(), ImmutableMap.of(
      "type", "hive",
      "default-namespace", "default",
      "parquet-enabled", "true",
      "cache-enabled", "false" // Spark will delete tables using v1, leaving the cache out of sync
  ));
```

2. new base abstract SparkSpecifyCatalogTestBase with three constructors

```
  public SparkSpecifyCatalogTestBase() {
    this(SparkCatalogType.TEST_HADOOP, null);
  }

  public SparkSpecifyCatalogTestBase(SparkCatalogType sparkCatalogType) {
    this(sparkCatalogType, null);
  }

  public SparkSpecifyCatalogTestBase(SparkCatalogType sparkCatalogType, Map<String, String> config) {
 .........
 }
```

Importantly, I'm not sure which test cases can be migrated to use it.

thx.

### CODE REVIEW

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/SparkRowLevelOperationsTestBase.java [hunk 1]
```diff
@@ -69,34 +70,34 @@ public SparkRowLevelOperationsTestBase(String catalogName, String implementation
...
-            true,
-            WRITE_DISTRIBUTION_MODE_NONE
+        {"testhive", SparkCatalog.class.getName(),
+         ImmutableMap.of(
+             "type", "hive",
+             "default-namespace", "default"
+         ),
+         "orc",
+         true,
+         WRITE_DISTRIBUTION_MODE_NONE
```
Comment (rdblue) [inline spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/SparkRowLevelOperationsTestBase.java:80]: Please remove all unnecessary formatting changes so that we can see just what changed. Thanks!

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  )),
+  TEST_HADOOP("testhadoop", SparkCatalog.class.getName(), ImmutableMap.of(
+      "type", "hadoop"
+  )),
+  SPARK_CATALOG("spark_catalog", SparkSessionCatalog.class.getName(), ImmutableMap.of(
+      "type", "hive",
+      "default-namespace", "default",
+      "parquet-enabled", "true",
+      "cache-enabled", "false" // Spark will delete tables using v1, leaving the cache out of sync
+  ));
```
Comment (kbendick) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java:38]: If we had constructors that provided a copy or something, so we had `SparkCatalogType.testHive(ImmutableMap.of("additionalConfig1", "additionalValue1")`, this would have much more benefit for the tests that need to override the `Parameters` to specify different catalog configuration.

Though that could be a v2 task.

**Review by kbendick — COMMENTED** [review]
Thanks for working on this @hililiwei! I just have one piece of input or opinion that I'd like to discuss (and hear other's opinions on if they care to chime in). Let a comment in the code on where that change _might_ be made.

In some of the tests that extend SparkCatalogTestBase, the catalogs are overridden by a new `@Parameters` section with only one or more catalogs. Often times, currently, this is done to allow for overriding the catalog `config`. But everything has to be redeclared.

Here's an example where we overrode the `@Parameters` to run a different set of catalogs (though still used all 3). The main reason for the override here was to allow for changing the catalog configuration with the additional stuff: https://github.com/apache/iceberg/blob/f5a753791f4dc6aca78569a14f731feda9edf462/spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkCatalogHadoopOverrides.java#L50-L77

If we could make it so that the config in the enum can be overridden without redeclaring everything (or whatever it winds up being if enums don't support that), that would be a big win in my view / really useful! This could be a task for a follow-up PR, but Very interested to hear other's opinions on this 🙂

File: spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/SparkRowLevelOperationsTestBase.java (hunk 1 above)
Comment (hililiwei) [reply spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/SparkRowLevelOperationsTestBase.java]: Done. In addition, I roll back the changes that were migrated to the new test  class from the last commit. Wait until the new test class is completed.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java (hunk 2 above)
Comment (hililiwei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java:38]: Thanks for your reply. 
Do you mean that the developers to specify only the additional config, without having to redeclare the entire catalog repeatedly?

SparkSpecifyCatalogTestBase provides the constructor as shown below, it can be used to add additional configuration items.

https://github.com/apache/iceberg/blob/9d431e87309275ecaf70c4a4b77f76a3dd921b90/spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java#L72-L79

Comment (hililiwei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java:38]: `super(SparkCatalogType.SPARK_CATALOG, ImmutableMap.of("additionalConfig1", "additionalValue1");`
Is it possible to use this method to achieve the desired results?

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  public String getCatalogName() {
+    return catalogName;
+  }
+
+  public String getImplementation() {
+    return implementation;
+  }
+
+  public Map<String, String> getConfig() {
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java:58]: Please remove `get` from method names.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark;
+
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+
+public enum SparkCatalogType {
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java:25]: I wouldn't say this is a catalog type, it's a catalog configuration.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  public SparkSpecifyCatalogTestBase(SparkCatalogType sparkCatalogType) {
+    this(sparkCatalogType, null);
+  }
+
+  public SparkSpecifyCatalogTestBase(SparkCatalogType sparkCatalogType, Map<String, String> config) {
+    this.implementation = sparkCatalogType.getImplementation();
+
+    this.catalogConfig = new HashMap<>(sparkCatalogType.getConfig());
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java:75]: We prefer using `Map.newHashMap`.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    if (config != null && !config.isEmpty()) {
+      config.forEach((key, value) -> catalogConfig.merge(key, value, (oldValue, newValue) -> newValue));
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java:77]: Why not just use `put`?

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestRuntimeFiltering.java [hunk 7]
```diff
@@ -44,10 +44,10 @@
 import static org.apache.spark.sql.functions.date_add;
 import static org.apache.spark.sql.functions.expr;
 
-public class TestRuntimeFiltering extends SparkCatalogTestBase {
+public class TestRuntimeFiltering extends SparkSpecifyCatalogTestBase {
 
-  public TestRuntimeFiltering(String catalogName, String implementation, Map<String, String> config) {
-    super(catalogName, implementation, config);
+  public TestRuntimeFiltering() {
+    super(SparkCatalogType.SPARK_CATALOG);
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestRuntimeFiltering.java:50]: I think we should use Hadoop for tests that only use one catalog. Also, if we're using Hadoop then we don't need to use the test base that sets up the metastore.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java (hunk 3 above)
Comment (hililiwei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java]: Done.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java (hunk 4 above)
Comment (hililiwei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogType.java]: > We could add more that have slightly different configs.
Could you provide more detailed guidance?I'm not familiar with this part.thank you

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestRuntimeFiltering.java (hunk 7 above)
Comment (hililiwei) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/source/TestRuntimeFiltering.java]: https://github.com/apache/iceberg/blob/f6fdeb0e63f3058eabf860bd893bea24f6649b4c/spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java#L54-L76

Can this be achieved by modifying the code as follows?

```

  @Before
  public void checkMetastoreAndSpark() {
    if (SparkTestBase.spark == null) {
      synchronized (SparkTestBase.class) {
        if (SparkTestBase.spark == null) {
          if (StringUtils.equals(catalogName, SPARK_CATALOG_HADOOP.catalogName())) {
            startSpark();
          } else {
            startMetastoreAndSpark();
          }
        }
      }
    }
  }

  public static void startSpark() {
    SparkTestBase.spark = SparkSession.builder()
        .master("local[2]")
        .config(SQLConf.PARTITION_OVERWRITE_MODE().key(), "dynamic")
        .getOrCreate();
  }

  public static void startMetastoreAndSpark() {
    SparkTestBase.metastore = new TestHiveMetastore();
    metastore.start();
    SparkTestBase.hiveConf = metastore.hiveConf();

    SparkTestBase.spark = SparkSession.builder()
        .master("local[2]")
        .config(SQLConf.PARTITION_OVERWRITE_MODE().key(), "dynamic")
        .config("spark.hadoop." + METASTOREURIS.varname, hiveConf.get(METASTOREURIS.varname))
        .enableHiveSupport()
        .getOrCreate();

    SparkTestBase.catalog = (HiveCatalog)
        CatalogUtil.loadCatalog(HiveCatalog.class.getName(), "hive", ImmutableMap.of(), hiveConf);

    try {
      catalog.createNamespace(Namespace.of("default"));
    } catch (AlreadyExistsException ignored) {
      // the default namespace already exists. ignore the create error
    }
  }

```

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.hadoop.HadoopCatalog;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+
+public abstract class SparkSpecifyCatalogTestBase extends SparkTestBase {
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkSpecifyCatalogTestBase.java:37]: How about `SparkSingleCatalogTestBase`? I think that is clear what this is for.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogConfig.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  SPARK_CATALOG_HADOOP("testhadoop", SparkCatalog.class.getName(), ImmutableMap.of(
+      "type", "hadoop"
+  )),
+  SPARK_SESSION_CATALOG_HIVE("spark_catalog", SparkSessionCatalog.class.getName(), ImmutableMap.of(
+      "type", "hive",
+      "default-namespace", "default",
+      "parquet-enabled", "true",
+      "cache-enabled", "false" // Spark will delete tables using v1, leaving the cache out of sync
+  )),
+  SPARK_SESSION_CATALOG_HADOOP("spark_catalog", SparkSessionCatalog.class.getName(), ImmutableMap.of(
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogConfig.java:39]: We only need to test Hive with the Spark session catalog. We don't recommend using other catalogs with the session catalog unless you really know what you're doing.

**Comment by rdblue** [comment]
@hililiwei, I opened a PR against your branch to make this a bit more generic and implement my suggestions.

**Comment by hililiwei** [comment]
> @hililiwei, I opened a PR against your branch to make this a bit more generic and implement my suggestions.

Thanks for your guidance. merged.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @hililiwei! This should really help us reduce test runtime in CI.
